app/models/ontology/mappings.rb

```diff
@@ -0,0 +1,86 @@
+module Ontology
+  module Mappings
+    extend ActiveSupport::Concern
+
+    included do
+      has_many :mappings,
+        autosave: false,
+        extend:   Methods
+    end
+
+    module Methods
+      def iri_for_child(*args)
+        proxy_association.owner.iri_for_child(*args)
+      end
+
+      def determine_mapping_type(typename)
+        raise 'mapping type missing' if typename.blank?
+        kind = Mapping::KINDS_MAPPING[typename]
+        if kind.nil?
+          key = Mapping::KINDS_MAPPING.keys.detect { |k| typename.include?(k) }
+          kind = Mapping::KINDS_MAPPING[key]
+        end
+        if kind.nil?
+          kind = Mapping::KINDS.detect { |k| typename.downcase.include?(k) }
+        end
+        kind || Mapping::DEFAULT_MAPPING_KIND
+      end
+
+      def update_or_create_from_hash(hash, user, timestamp = Time.now)
```

Comment:
Lint/UnusedMethodArgument: Unused method argument - `user`. If it's necessary, use `_` or `_user` as an argument name to indicate that it won't be used.<br>Metrics/CyclomaticComplexity: Cyclomatic complexity for update_or_create_from_hash is too high. [10/6]<br>Metrics/MethodLength: Method has too many lines. [43/10]<br>Metrics/PerceivedComplexity: Perceived complexity for update_or_create_from_hash is too high. [10/7]